crates/circuit/src/slice.rs

```diff
@@ -0,0 +1,375 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use thiserror::Error;
+
+use pyo3::exceptions::PyIndexError;
+use pyo3::prelude::*;
+use pyo3::types::PySlice;
+
+use self::sealed::{Descending, SequenceIndexIter};
+
+/// A Python-space indexer for the standard `PySequence` type; a single integer or a slice.
+///
+/// These come in as `isize`s from Python space, since Python typically allows negative indices.
+/// Use `with_len` to specialize the index to a valid Rust-space indexer into a collection of the
+/// given length.
+pub enum PySequenceIndex<'py> {
+    Int(isize),
+    Slice(Bound<'py, PySlice>),
+}
+
+impl<'py> FromPyObject<'py> for PySequenceIndex<'py> {
+    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
+        // `slice` can't be subclassed in Python, so it's safe (and faster) to check for it exactly.
+        // The `downcast_exact` check is just a pointer comparison, so while `slice` is the less
+        // common input, doing that first has little-to-no impact on the speed of the `isize` path,
+        // while the reverse makes `slice` inputs significantly slower.
+        if let Ok(slice) = ob.downcast_exact::<PySlice>() {
+            return Ok(Self::Slice(slice.clone()));
+        }
+        Ok(Self::Int(ob.extract()?))
```

Comment:
Heh, this makes me think of: https://github.com/Qiskit/rustworkx/pull/1096 where you opted for just flipping the order. This is starting to make me think maybe we want to contribute this to PyO3 so there is a common impl of this for everyone.

Reply:
You're just watching me learning more of the low-level internals of everything live haha.

In Rustworkx I didn't know enough other than to leave the `FromPyObject` derivation in place, in which case flipping the order just makes things faster.  But for the sake of completeness here, since I'd already wasted enough time that a little longer made no difference, I actually checked whether `slice` is subclassable (turns out it isn't), in which case the derived `FromPyObject` is needlessly slow; you can't subclass it, so the `type is slice` check is sufficient and is just a pointer comparison.